### PR TITLE
Feature/physic manager

### PR DESCRIPTION
**Pull request by SomeOne768**

### CODE REVIEW

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/architecture/Time.hpp [hunk 1]
```diff
@@ -12,17 +12,18 @@
...
  * This class is a singleton.
  */
 class Time
 {
-public:
+private:
 	/**
 	 * Instance of the singleton.
 	 */
 	static Time* instance;
```
Comment (SomeOne768) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/architecture/Time.hpp:24]: better to put it private, to ensure encapsulation

Comment (SomeOne768) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/architecture/Time.hpp:24]: better to put it private, to ensure encapsulation

**Comment by SomeOne768** [comment]
Refacto  + architecture to update 
Need tests and time measurement after to compare gain.

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/CMakeLists.txt [hunk 2]
```diff
@@ -19,6 +19,8 @@ set(_SAINT_NEC_ENGINE_LIB_SRC
 	"physics/PhysicObject.cpp"
 	"physics/util.cpp"
 	"graphics/TessellatedPlane.cpp"
+	"architecture/PhysicManager.cpp"
+	"physics/algorithms.cpp"
```
Comment (antonindevidal) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/CMakeLists.txt:23]: Lower case first letter

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/BoxCollider.cpp [hunk 3]
```diff
@@ -98,4 +98,20 @@ namespace sne::saintNecPhysics
     {
         return hasSATCollision(*this, b);
     }
+
+    bool BoxCollider::intersection(const Collider *c,const glm::vec3 &axis) const 
+    {
+        return c->intersection(*this, axis);
+    }
+
+    #include "util.hpp"
```
Comment (antonindevidal) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/BoxCollider.cpp]: include in hpp

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/algorithms.cpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+                      return p1->getPosition()[0] < p2->getPosition()[0]; // to update for min point axis
+                  });
+
+        glm::vec3 axis{1, 0, 0};
+        std::vector<PhysicObject *> curr;
+
+        for (int i = 0; i < v.size(); i++)
+        {
+            // For each object
+            for (std::vector<PhysicObject *>::iterator elt = curr.begin(); elt != curr.end(); elt++)
```
Comment (antonindevidal) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/algorithms.cpp:25]: I don't know if it is voluntary but here it copies an entire array each time

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/util.cpp [hunk 5]
```diff
@@ -1,7 +1,80 @@
...
+            double boxe1_projection = dot(axis, boxe1[i]),
+                   boxe2_projection = dot(axis, boxe2[i]);
+
+            boxe1_min = (boxe1_min < boxe1_projection) ? boxe1_min : boxe1_projection;
+            boxe1_max = (boxe1_max > boxe1_projection) ? boxe1_max : boxe1_projection;
+            boxe2_min = (boxe2_min < boxe2_projection) ? boxe2_min : boxe2_projection;
+            boxe2_max = (boxe2_max > boxe2_projection) ? boxe2_max : boxe2_projection;
+        }
+
+        if (areDisjoint(boxe1_min, boxe1_max, boxe2_min, boxe2_max))
```
Comment (antonindevidal) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/util.cpp]: Can be rafactored in return areDisjoint...)

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/util.cpp [hunk 6]
```diff
@@ -1,7 +1,80 @@
...
+
+        for (int i = 1; i < boxe.getNbPoints(); i++)
+        {
+            double boxe_projection = dot(axis, boxe[i]);
+
+            boxe_min = (boxe_min < boxe_projection) ? boxe_min : boxe_projection;
+            boxe_max = (boxe_max > boxe_projection) ? boxe_max : boxe_projection;
+        }
+
+        if (areDisjoint(boxe_min, boxe_max, sphere_min, sphere_max))
```
Comment (antonindevidal) [inline Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/util.cpp]: same

**Review by antonindevidal — APPROVED** [review]

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/CMakeLists.txt (hunk 2 above)
Comment (SomeOne768) [reply Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/CMakeLists.txt:23]: As it is not a class but only function shouldn't we keep lower case like for 'util' ?

File: Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/algorithms.cpp (hunk 4 above)
Comment (SomeOne768) [reply Saint-Nec-Engine/Saint-Nec-Engine-Lib/src/physics/algorithms.cpp:25]: Here we use iterator (pointer) so, if I am not wrong there is no copy

**Pull request merged**